File: archive/python-legacy/polymarket/data_pipeline/nws_fetcher.py
"""
National Weather Service (NWS) data fetcher module
Retrieves weather observations from api.weather.gov
Free, no API key required, US only
"""

import logging
import requests
import pandas as pd
from datetime import datetime, timedelta
from typing import List, Dict, Optional

logger = logging.getLogger(__name__)

NWS_STATIONS = {
    "NYC": {"station_id": "KNYC", "name": "New York City, NY"},
    "LA": {"station_id": "KLAX", "name": "Los Angeles, CA"},
    "Chicago": {"station_id": "KORD", "name": "Chicago O'Hare, IL"},
    "Dallas": {"station_id": "KDFW", "name": "Dallas/Fort Worth, TX"},
    "Denver": {"station_id": "KDEN", "name": "Denver, CO"},
    "Miami": {"station_id": "KMIA", "name": "Miami, FL"},
    "Boston": {"station_id": "KBOS", "name": "Boston, MA"},
}

BASE_URL = "https://api.weather.gov"


class NWSFetcher:
    """Fetches weather observations from National Weather Service API"""

    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            "User-Agent": "PolymarketWeatherPredictor/1.0 (weather-prediction@example.com)",
            "Accept": "application/geo+json",
        })

    def fetch_daily_observations(
        self,
        station_id: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """
        Fetch daily weather observations for a station.
        NWS API returns hourly observations; we aggregate to daily.

        Args:
            station_id: NWS station ID (e.g., "KNYC")
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)

        Returns:
            DataFrame with daily aggregated weather observations
        """
        start = datetime.strptime(start_date, "%Y-%m-%d")
        end = datetime.strptime(end_date, "%Y-%m-%d")

        all_observations = []
        # NWS API limits to 500 observations per request; fetch in weekly chunks
        current = start
        while current < end:
            chunk_end = min(current + timedelta(days=7), end)
            obs = self._fetch_observations_chunk(station_id, current, chunk_end)
            all_observations.extend(obs)
            current = chunk_end

        if not all_observations:
            logger.warning(f"NWS: No observations for {station_id}")
            return pd.DataFrame()

        df = pd.DataFrame(all_observations)
        daily = self._aggregate_to_daily(df, station_id)

        logger.info(f"NWS: Fetched {len(daily)} daily observations for {station_id}")
        return daily

    def _fetch_observations_chunk(
        self,
        station_id: str,
        start: datetime,
        end: datetime,
    ) -> List[Dict]:
        """Fetch a chunk of observations from NWS API"""
        url = f"{BASE_URL}/stations/{station_id}/observations"
        params = {
            "start": start.strftime("%Y-%m-%dT00:00:00Z"),
            "end": end.strftime("%Y-%m-%dT23:59:59Z"),
        }

        try:
            response = self.session.get(url, params=params, timeout=30)
            response.raise_for_status()
            data = response.json()

            observations = []
            for feature in data.get("features", []):
                props = feature.get("properties", {})
                temp = props.get("temperature", {}).get("value")
                wind = props.get("windSpeed", {}).get("value")
                precip = props.get("precipitationLastHour", {}).get("value")
                timestamp = props.get("timestamp")

                if timestamp:
                    observations.append({
                        "timestamp": timestamp,
                        "temperature": temp,
                        "wind_speed": wind / 3.6 if wind is not None else None,  # km/h to m/s
                        "precipitation": precip if precip is not None else 0,
                    })

            return observations

        except requests.RequestException as e:
            logger.error(f"NWS error for {station_id}: {e}")
            return []

    def _aggregate_to_daily(self, df: pd.DataFrame, station_id: str) -> pd.DataFrame:
        """Aggregate hourly observations to daily summaries"""
        df["timestamp"] = pd.to_datetime(df["timestamp"])
        df["date"] = df["timestamp"].dt.date

        daily = df.groupby("date").agg(
            temperature_max=("temperature", "max"),
            temperature_min=("temperature", "min"),
            temperature_mean=("temperature", "mean"),
            precipitation_total=("precipitation", "sum"),
            wind_speed_mean=("wind_speed", "mean"),
        ).reset_index()

        daily["date"] = pd.to_datetime(daily["date"])
        daily["station_id"] = f"NWS_{station_id}"

        return daily

    def fetch_location(
        self,
        location_key: str,
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for a pre-configured location"""
        station = NWS_STATIONS.get(location_key)
        if not station:
            logger.warning(f"NWS: No station for {location_key} (US only)")
            return pd.DataFrame()

        return self.fetch_daily_observations(
            station_id=station["station_id"],
            start_date=start_date,
            end_date=end_date,
        )

    def fetch_multiple_locations(
        self,
        location_keys: List[str],
        start_date: str,
        end_date: str,
    ) -> pd.DataFrame:
        """Fetch data for multiple locations and combine"""
        all_data = []
        for key in location_keys:
            df = self.fetch_location(key, start_date, end_date)
            if not df.empty:
                all_data.append(df)

        if all_data:
            return pd.concat(all_data, ignore_index=True)
        return pd.DataFrame()

File: src/data_pipeline/nws_fetcher.rs
use std::collections::HashMap;

use chrono::{DateTime, Duration, NaiveDate, Utc};
use serde_json::Value;

use crate::types::WeatherRecord;

const BASE_URL: &str = "https://api.weather.gov";

// Type aliases for complex observation types
type Observation = (DateTime<Utc>, Option<f64>, Option<f64>, Option<f64>);
type DailyData = (Option<f64>, Option<f64>, Option<f64>);

#[derive(Clone)]
pub struct NWSFetcher {
    client: reqwest::blocking::Client,
    stations: HashMap<String, String>,
}

impl NWSFetcher {
    pub fn new() -> Self {
        Self {
            client: crate::data_pipeline::build_client(30),
            stations: default_stations(),
        }
    }

    pub fn fetch_daily_observations(
        &self,
        station_id: &str,
        start_date: NaiveDate,
        end_date: NaiveDate,
    ) -> Vec<WeatherRecord> {
        let mut all_observations: Vec<Observation> = Vec::new();

        let mut current = start_date;
        while current < end_date {
            let chunk_end = (current + Duration::days(7)).min(end_date);
            let mut chunk = self.fetch_observations_chunk(station_id, current, chunk_end);
            all_observations.append(&mut chunk);
            current = chunk_end;
        }

        self.aggregate_to_daily(all_observations, station_id)
    }

    fn fetch_observations_chunk(
        &self,
        station_id: &str,
        start: NaiveDate,
        end: NaiveDate,
    ) -> Vec<Observation> {
        let url = format!("{BASE_URL}/stations/{station_id}/observations");
        let query = [
            (
                "start",
                format!("{}T00:00:00Z", start.format("%Y-%m-%d")),
            ),
            (
                "end",
                format!("{}T23:59:59Z", end.format("%Y-%m-%d")),
            ),
        ];

        let Ok(resp) = self.client.get(url).query(&query).send() else {
            return Vec::new();
        };
        let Ok(json) = resp.json::<Value>() else {
            return Vec::new();
        };

        let mut out = Vec::new();
        for feature in json
            .get("features")
            .and_then(|f| f.as_array())
            .cloned()
            .unwrap_or_default()
        {
            let props = feature.get("properties").cloned().unwrap_or(Value::Null);
            let ts = props.get("timestamp").and_then(|v| v.as_str());
            let Some(ts) = ts else {
                continue;
            };
            let Ok(dt) = DateTime::parse_from_rfc3339(ts) else {
                continue;
            };
            let dt = dt.with_timezone(&Utc);
            let temp = props
                .get("temperature")
                .and_then(|v| v.get("value"))
                .and_then(|v| v.as_f64());
            let wind = props
                .get("windSpeed")
                .and_then(|v| v.get("value"))
                .and_then(|v| v.as_f64())
                .map(|x| x / 3.6);
            let precip = props
                .get("precipitationLastHour")
                .and_then(|v| v.get("value"))
                .and_then(|v| v.as_f64())
                .or(Some(0.0));

            out.push((dt, temp, wind, precip));
        }

        out
    }

    fn aggregate_to_daily(
        &self,
        observations: Vec<Observation>,
        station_id: &str,
    ) -> Vec<WeatherRecord> {
        let mut by_day: HashMap<NaiveDate, Vec<DailyData>> = HashMap::new();
        for (dt, temp, wind, precip) in observations {
            by_day.entry(dt.date_naive()).or_default().push((temp, wind, precip));
        }

        let mut days: Vec<NaiveDate> = by_day.keys().copied().collect();
        days.sort();

        let mut out = Vec::new();
        for day in days {
            let rows = by_day.remove(&day).unwrap_or_default();
            let temps: Vec<f64> = rows.iter().filter_map(|(t, _, _)| *t).collect();
            let winds: Vec<f64> = rows.iter().filter_map(|(_, w, _)| *w).collect();
            let precips: Vec<f64> = rows.iter().filter_map(|(_, _, p)| *p).collect();

            let mut rec = WeatherRecord::new(day, format!("NWS_{station_id}"));
            rec.temperature_max = temps.iter().copied().reduce(f64::max);
            rec.temperature_min = temps.iter().copied().reduce(f64::min);
            rec.temperature_mean = if temps.is_empty() {
                None
            } else {
                Some(temps.iter().sum::<f64>() / temps.len() as f64)
            };
            rec.precipitation_total = Some(precips.iter().sum());
            rec.wind_speed_mean = if winds.is_empty() {
                None
            } else {
                Some(winds.iter().sum::<f64>() / winds.len() as f64)
            };
            out.push(rec);
        }

        out
    }

    pub fn fetch_location(&self, location_key: &str, start_date: NaiveDate, end_date: NaiveDate) -> Vec<WeatherRecord> {
        let Some(station_id) = self.stations.get(location_key) else {
            return Vec::new();
        };
        self.fetch_daily_observations(station_id, start_date, end_date)
    }
}

impl Default for NWSFetcher {
    fn default() -> Self {
        Self::new()
    }
}

fn default_stations() -> HashMap<String, String> {
    [
        ("NYC", "KNYC"),
        ("LA", "KLAX"),
        ("Chicago", "KORD"),
        ("Dallas", "KDFW"),
        ("Denver", "KDEN"),
        ("Miami", "KMIA"),
        ("Boston", "KBOS"),
    ]
    .into_iter()
    .map(|(k, v)| (k.to_string(), v.to_string()))
    .collect()
}
